codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+                    // TODO: swap to failing the promise. Today, however, the promise is likely
+                    // being ignored.
+                    throw new IllegalArgumentException(
+                        "Http2StreamFrame.stream() must be an Http2StreamChannel: " + streamObject);
+                }
+                int streamId = ((Http2StreamChannel) streamObject).stream.id();
+                if (msg instanceof Http2DataFrame) {
+                    Http2DataFrame frame = (Http2DataFrame) msg;
+                    http2Handler.encoder().writeData(http2HandlerCtx, streamId, frame.content().retain(),
+                        frame.padding(), frame.endStream(), promise);
+                } else if (msg instanceof Http2HeadersFrame) {
+                    Http2HeadersFrame frame = (Http2HeadersFrame) msg;
+                    http2Handler.encoder().writeHeaders(
+                        http2HandlerCtx, streamId, frame.headers(), frame.padding(), frame.endStream(), promise);
+                } else if (msg instanceof Http2ResetFrame) {
+                    Http2ResetFrame frame = (Http2ResetFrame) msg;
+                    http2Handler.resetStream(http2HandlerCtx, streamId, frame.errorCode(), promise);
+                } else {
+                    unexpected = true;
+                }
+            } else if (msg instanceof Http2GoAwayFrame) {
+                Http2GoAwayFrame frame = (Http2GoAwayFrame) msg;
+                int lastStreamId = http2Handler.connection().remote().lastStreamCreated()
+                    + frame.extraStreamIds() * 2;
+                http2Handler.goAway(
+                    http2HandlerCtx, lastStreamId, frame.errorCode(), frame.content().retain(), promise);
+            } else {
+                unexpected = true;
+            }
+            if (unexpected) {
+                // TODO: swap to failing the promise. Today, however, the promise is likely being
+                // ignored.
+                throw new IllegalStateException("Unexpected Http2Frame type: " + msg.getClass().getName());
+            }
+        } finally {
+            ReferenceCountUtil.release(msg);
+        }
+    }
+
+    Http2StreamChannel createStreamChannel(ChannelHandlerContext ctx, Http2Stream stream,
+            ChannelHandler handler) {
+        EventLoopGroup group = streamGroup != null ? streamGroup : ctx.channel().eventLoop();
+        Http2StreamChannel channel = new Http2StreamChannel(stream);
+        channel.pipeline().addLast(handler);
+        group.register(channel);
+        return channel;
+    }
+
+    private static ChannelHandlerInvoker invoker(ChannelHandlerContext ctx) {
+        return ((EventLoop) ctx.executor()).asInvoker();
+    }
+
+    /**
+     * Notifies any child streams of the read completion.
+     */
+    @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) {
+        for (int i = 0; i < streamsToFireChildReadComplete.size(); i++) {
```

Comment:
I think this is a continuation of the discussion above: is it reasonable to assume `ctx.fire*` methods don't throw? If it isn't, then there will be a lot of pain to handle the error cases or (more likely) I will make my own set of methods that behave similarly to `ChannelHandlerInvokerUtil` but call the `ctx.fire*` methods.